### ISSUE TITLE
[BUG] Freezing solr queue worker

### ISSUE THREAD

**Issue by jackd248** [TYPO3-Solr/ext-solr#2500]
**Describe the bug**
The scheduler task "Solr Queue Worker" freezes sometimes while indexing a large amount of items. The process is stucked and runs forever. When stopping the task manually and restarting it again, everything works fine and without problems. But with this behaviour the automatic indexing of new items isn't working anymore. 

The EXT:solr module in the TYPO3 backend gives no sign of a problem. Only the `solr.log` shows me the following hint:

`Failed indexing page Index Queue item: 176920`

This appears several times and I can only assume a relation.

**Expected behavior**
All queue items should be indexed as aspected.

**Used versions (please complete the following information):**
- TYPO3 Version: 9.5.11
- EXT:solr Version: 10.0.1
- Used Apache Solr Version: 7.6.0
- PHP Version: 7.2.23

**Comment by dkd-kaehm** [comment]
> The scheduler task "Solr Queue Worker" freezes sometimes while indexing a large amount of items.

Do you know about "Number of documents to index"-setting in "Index Queue Worker (solr)" Task?

---

Please check the system processes during freezes of queue. I think you'll find some running processes(zomby?). 
We should find the cause of the crushes.(may be memory_limit, infinite loop, sql timeouts, etc.)
We should check how we can mark the Indexing Queue Task as not running if php process crushes(may be PHPs shutdown callback).

**Comment by jackd248** [comment]
Yes, I know about this. Currently 20 items are processed every 5 mins. I thought this would be low enough, isn't it? (we have about 2000 items in the index queue)

**Comment by sfsmfc** [comment]
@jackd248 Can you have a look into your solr server interface for some server error. We have a similar problem and this causes from `out of memory` while indexing.

**Comment by jackd248** [comment]
@sfsmfc Unfortunately the solr server interface looks pretty clean so far. Do you found the error only in the interface or do I have to check a server log as well?

**Comment by sfsmfc** [comment]
Should be also in log.

**Comment by jackd248** [comment]
During development we jumped from EXT:solr version 9 to 10. And because of this is now a live system it's not so easy to upgrade the Solr server itself to the newest (and required) version 8.2. 
I'm not sure, if this could cause the issue.

**Comment by dmitryd** [comment]
This typically happens if you run out of memory. Solr loads all indexing records for your specified amount into RAM and if you specify a lot (>1000 according to my experience), then it may fail with out of memory (I have a 256M limit).

Lower your limit. There is nothing that solr can do here.

**Comment by jackd248** [comment]
I don't think it's a problem with the memory, because the problem also appears while the "daily reindexing" of items. So only a few entities (almost exclusively event and news items) are stucked in the queue.

**Comment by dmitryd** [comment]
@jackd248 Search typo3 logs for exceptions. They would also leave the job stuck. File name is usually like `typo3_*.log`.

**Comment by dkd-kaehm** [comment]
@jackd248 
Some news on that issue?

A litle explanation about the trouble: 
The scheduler task of TYPO3 is not marked as "not running", because the process is frozen or killed before the method for "marking the tasks as ended" can be called. 
So the task is always marked as "currently running" and will be newer executed by TYPO3 Scheduler. 
And we want to detect the cause of that.

There are two options:
A. The PHP process is killed by some "OS cleanup task" or max_execution_time or whatever.
B. There is infinite loop and the PHP process stucks and doesn't ever come to the method for "marking the tasks as ended"

We want more infos about:
1. What causes that behaviour the A or the B from above?
1.1. Scheduler execution context: CLI or WEB?
1.2. 100% assurance about "nobody kills the Scheduler PHP process"
      You told, that that happens "while indexing a large amount of items", which is the sign for a longer execution times...
2 Your supposition or statement about that trouble.

If You can't assure, that nobody kills the PHP process, there is currently no sign for infinite loop trouble, and I would propose to close that issue.

---

This issue will be be closed automatically after 28 days of silence.

**Comment by dkd-kaehm** [comment]
Closing because of silence.

**Comment by Raidri** [comment]
Hi,
I want to bring back this issue.
I've the same problem.

We use TYPO3 11 with solr 11.5.1 both are hosted by Mittwald.

The system use 18 languages and we want to index data from a custom extension model.

In our case the scheduler process never stops, it seems to be frozen and keep running in the backend scheduler view. But I don't get any errors.
the index queue worker will started by the CLI (cronjob) every 15 minutes without parallel execution.

I started to set the amount of items to index by 100. If I understand solr correct that's 100 * 18 items because of the 18 languages. After ~25 Items the scheduler is frozen and no further items will be added to solr.

I have to reduce the items which should be indexed to 25. Then it works.

**Comment by dkd-kaehm** [comment]
@Raidri 
can you see something in PHPs system error log about memory_limit?

**Comment by dmitryd** [comment]
It would be good to know what are memory settings for this site. In my experiences all such "task was stuck" cases were "out of memory" issues. But I was under impression that TYPO3 is supposed to detect that the task was abnormally terminated and unmark it as running.

**Comment by Raidri** [comment]
@dkd-kaehm and @dmitryd that was a great hint for me. It's true, I running in a memory_leak. 
Sorry for reopen the issue. I didn't except a memory leak.

Thanks for your help :)

**Comment by dkd-kaehm** [comment]
@Raidri 
It would be nice if you find exact cause of memory-guzzler and the source of trouble.
May be something like that is the causer: 
https://typo3.slack.com/archives/C02FF05Q4/p1671749002691819
